incubator/hnc/pkg/controllers/config_controller.go

```diff
@@ -91,6 +116,63 @@ func (r *ConfigReconciler) writeSingleton(ctx context.Context, inst *api.HNCConf
 	return nil
 }
 
+// createObjectReconcilers creates corresponding ObjectReconcilers for the newly added types in the
+// HNC configuration, if there is any. It also informs the in-memory forest about the newly created
+// ObjectReconcilers.
+func (r *ConfigReconciler) createObjectReconcilers(inst *api.HNCConfiguration) error {
+	// This method is guarded by the forest mutex. The mutex is guarding two actions: creating ObjectReconcilers for
+	// the newly added types and adding the created ObjectReconcilers to the types list in the Forest (r.Forest.types).
+	//
+	// We use mutex to guard write access to the types list because the list is a shared resource between various
+	// reconcilers. It is sufficient because both write and read access to the list are guarded by the same mutex.
+	//
+	// We use the forest mutex to guard ObjectReconciler creation together with types list modification so that the
+	// forest cannot be changed by the HierarchyReconciler until both actions are finished. If we do not use the
+	// forest mutex to guard both actions, HierarchyReconciler might change the forest structure and read the types list
+	// from the forest for the object reconciliation, after we create ObjectReconcilers but before we write the
+	// ObjectReconcilers to the types list. As a result, objects of the newly added types might not be propagated correctly
+	// using the latest forest structure.
+	r.Forest.Lock()
+	defer r.Forest.Unlock()
+	for _, t := range inst.Spec.Types {
+		gvk := schema.FromAPIVersionAndKind(t.APIVersion, t.Kind)
+		if r.Forest.HasTypeSyncer(gvk) {
+			continue
+		}
+		if err := r.createObjectReconciler(gvk); err != nil {
+			r.Log.Error(err, "while creating an object controller")
+			return err
```

Comment:
Generally, you should _either_ log _or_ return an error. If an upstream caller needs to know there was an error, you add information to the error (eg `return fmt.Errorf("while trying to foo: %s", err)`); if the caller doesn't need to know, log it and don't return anything.

Otherwise, one actual error ends up causing N errors in the log, where N is the call stack depth of the code in question.

Reply:
Thanks so much for the explanation. In this case, I changed to return the error without logging. The Reconcile() function will log the error.